cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkTransactionService.java

```diff
@@ -0,0 +1,314 @@
+        return;
+      }
+
+      // Get the transaction
+      JobTransaction jobTransaction = jobTransactions.get(jobId);
+      if (jobTransaction == null) {
+        // The only reason we can find the jobId from the stageToJob map but not the job transaction is because
+        // the job is completed, hence the transaction get removed. In normal case, it shouldn't happen
+        // as a job won't complete if there are still stages running and
+        // this method only gets called from stage running in executor node.
+        responder.sendString(HttpResponseStatus.GONE,
+                             "No transaction associated with the stage " + stageId + " of job " + jobId);
+        return;
+      }
+
+      Transaction transaction = jobTransaction.getTransaction();
+      if (transaction == null) {
+        // Job failed to start a transaction. Response with GONE as well so that the stage execution can fail itself
+        responder.sendString(HttpResponseStatus.GONE,
+                             "Failed to start transaction for stage " + stageId + " of job " + jobId);
+        return;
+      }
+
+      // Serialize the transaction and send it back
+      try {
+        responder.sendByteArray(HttpResponseStatus.OK, txCodec.encode(transaction), null);
+      } catch (IOException e) {
+        // Shouldn't happen
+        LOG.error("Failed to encode Transaction {}", jobTransaction, e);
+        responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR,
+                             "Failed to encode transaction: " + e.getMessage());
+      }
+    }
+  }
+
+  /**
+   * A private class for handling the {@link Transaction} lifecycle for a job.
+   */
+  private final class JobTransaction {
+    private final Integer jobId;
+    private final Set<Integer> stageIds;
+    private final boolean ownTransaction;
+    private volatile Optional<Transaction> transaction;
+
+    JobTransaction(Integer jobId, Set<Integer> stageIds, @Nullable Transaction transaction) {
+      this.jobId = jobId;
+      this.stageIds = ImmutableSet.copyOf(stageIds);
+      this.ownTransaction = transaction == null;
+      this.transaction = transaction == null ? null : Optional.of(transaction);
+    }
+
+    Integer getJobId() {
+      return jobId;
+    }
+
+    Set<Integer> getStageIds() {
+      return stageIds;
+    }
+
+    /**
+     * Returns the {@link Transaction} associated with the job. If transaction hasn't been started, a new long
+     * transaction will be started.
+     *
+     * @return the job's {@link Transaction} or {@code null} if it failed to start transaction for the job.
+     */
+    @Nullable
+    public Transaction getTransaction() {
+      Optional<Transaction> tx = transaction;
+      if (tx == null) {
+        // double-checked locking
+        synchronized (this) {
+          tx = transaction;
+          if (tx == null) {
+            try {
+              tx = transaction = Optional.of(txClient.startLong());
```

Comment:
I think the main difference between long and short tx, besides it can get timeout or not, is the conflict detection done by the tx server. However, I don't think Tephra currently supports conflict detection when writes are distributed, since it doesn't support appending to change set. That is why I choose to use long transaction here. Correct me if I am wrong.